algebird-core/src/main/scala/com/twitter/algebird/Preparer.scala

```diff
@@ -0,0 +1,235 @@
+    val (a1, a2) = fn(MapPreparer.identity[T])
+    aggregate(a1.join(a2))
+  }
+
+  /**
+   * head and following methods all just call aggregate with one of the standard Aggregators.
+   * see the Aggregator object for more docs.
+   */
+
+  def head = aggregate(Aggregator.head)
+  def last = aggregate(Aggregator.last)
+  def max(implicit ord: Ordering[T]) = aggregate(Aggregator.max)
+  def maxBy[U: Ordering](fn: T => U) = {
+    implicit val ordT = Ordering.by(fn)
+    aggregate(Aggregator.max[T])
+  }
+
+  def min(implicit ord: Ordering[T]) = aggregate(Aggregator.min)
+  def minBy[U: Ordering](fn: T => U) = {
+    implicit val ordT = Ordering.by(fn)
+    aggregate(Aggregator.min[T])
+  }
+
+  def sum(implicit sg: Semigroup[T]) = aggregate(Aggregator.fromSemigroup(sg))
+  def reduce(fn: (T, T) => T) = aggregate(Aggregator.fromReduce(fn))
+}
+
+object MapPreparer {
+  /**
+   * Create a concrete MapPreparer.
+   */
+  def apply[A, T](fn: A => T) = new MapPreparer[A, T] { val prepareFn = fn }
+
+  /**
+   * This is purely an optimization for the case of mapping by identity.
+   * It overrides the key methods to not actually use the identity function.
+   */
+  def identity[A] = new MapPreparer[A, A] {
+    val prepareFn = (a: A) => a
+    override def map[U](fn: A => U) = MapPreparer(fn)
+    override def flatMap[U](fn: A => TraversableOnce[U]) = FlatMapPreparer(fn)
+    override def monoidAggregate[B, C](aggregator: MonoidAggregator[A, B, C]) = aggregator
+    override def aggregate[B, C](aggregator: Aggregator[A, B, C]) = aggregator
+  }
+}
+
+/**
+ * A Preparer that has had one or more flatMap operations applied.
+ * It can only accept MonoidAggregators.
+ */
+trait FlatMapPreparer[A, T] extends Preparer[A, T] {
+
+  def prepareFn: A => TraversableOnce[T]
+
+  def map[U](fn: T => U): FlatMapPreparer[A, U] =
+    FlatMapPreparer { a: A => prepareFn(a).map(fn) }
+
+  def flatMap[U](fn: T => TraversableOnce[U]) =
+    FlatMapPreparer { a: A => prepareFn(a).flatMap(fn) }
+
+  def monoidAggregate[B, C](aggregator: MonoidAggregator[T, B, C]): MonoidAggregator[A, B, C] =
+    aggregator.sumBefore.composePrepare(prepareFn)
+
+  /**
+   * alias of monoidAggregate for convenience
+   * unlike MapPreparer's aggregate, can only take MonoidAggregator
+   */
+  def aggregate[B, C](aggregator: MonoidAggregator[T, B, C]) = monoidAggregate(aggregator)
+
+  /**
+   * Like monoidAggregate, but using an implicit Monoid to construct the Aggregator
+   */
+  def sum(implicit monoid: Monoid[T]) = monoidAggregate(Aggregator.fromMonoid(monoid))
+
+  /**
+   * Split the processing into two parallel aggregations.
+   * You provide a function which produces two different aggregators from this preparer,
+   * and it will return a single aggregator which does both aggregations in parallel.
+   * (See also Aggregator's join method.)
+   *
+   * We really need to generate N versions of this for 3-way, 4-way etc splits.
+   */
```

Comment:
It's irritating that this returns an Aggregator rather than a MonoidAggregator. See issue #403.